enhancements/etcd/etcd-size-tuning.md

```diff
@@ -0,0 +1,230 @@
+* The backend limit can only be increased and not decreased.
+
+### Non-Goals
+
+* Handle consuming value changes without an etcd rollout.
+
+## Proposal
+
+Allow the customer/admin access to set the backend quota directly through a human readable value via the etcd.operator.openshift.io/v1 techpreview field. The only restriction is that they are disallowed from setting the value lower than the current.
+
+Increasing this value will not have etcd immediately reserve this space on disk, it just allows etcd to grow up to that value if etcd grows under the usage of the cluster.
+
+The default value or unset value would be the current, hardcoded backend quota: 8GiB. This would allow for upgrades where this configuration value wouldn't nesscessarily be set.
+An initial upper limit of 32GiB will be enforced so as to limit the impact of some etcd operations, like restore, transporting snapshots, etc. 32GiB is an arbitrary value and work with the Perf-Scale team will need to be done to validate this value and potentially increase the upper limit; finding a reasonable and safe upper bound will need to be done before this feature is made GA.
+
+We will need to work with the perf team to understand the memory requirements of largers quotas, and the effect of the quota on things like defragmentation, snapshot restore, etc.
+A downside is that with the profiles, discussed in the Alternatives section, there are a discrete, small, number of permutations to test, giving more confidence of the exact effects a given profile has on performance/memory requirements.
+This is a minor consideration as setting the quota to a arbitrarily higher value does not mean that etcd will take up that much space; it will depend on the workload.
+
+### Workflow Description
+
+**cluster administrator** is a human user responsible deploying and maintaining the cluster.
+
+1. The cluster administrator decides to change the etcd backend quota from the default (8GiB) to a larger value; e.g. 20GiB.
+2. They set the new value via the `etcd.operator.openshift.io/v1` API.
+3. An etcd redeployment will be automatically run which restarts the etcd pods which consume the new value.
+
+#### Invalid value
+
+1. The cluster administrator wants to set the quota from a larger value (e.g. 32GiB) to a smaller one (8GiB).
+2. They attempt to set the new value via the `etcd.operator.openshift.io/v1` API.
+3. The API server should return an error notifying them that the value is invalid - not allowed to shrink the quota.
+
+### API Extensions
+
+In operator/v1/0000_12_etcd-operator_01_config-TechPreviewNoUpgrade.crd.yaml:
+- Add BackendQuotaGiB of type int64
+  - With default '8GiB'.
+  - With validation disallow a decrease in value.
+  - With an upper limit of '32GiB'.
+
+```go
+type EtcdSpec struct {
+  // BackendQuotaGiB sets the etcd backend storage size limit in gibibytes. Defaults to 8GiB.
+  // +kubebuilder:validation:Optional
+  // +kubebuilder:default=8
```

Comment:
I have a case to argue about the default value tag 

- you can find in the default [test](blob:https://prow.ci.openshift.org/009d7cee-20a8-4c1f-9ba4-be4f99a8314c), that expects the `EtcdSpec` to be empty, it expects  `"backendQuotaGiB": <int64>8` 
- this is due to the fact that the field is not a pointer
- since the field is `int64` it should defaults to `0`
- however, because of this tag, it defaults to `8`
- I propose to replace the default tag with ` +kubebuilder:validation:Minimum=8` to avoid these issues
- However, as I have initially proposed a `pointer` field would have been more suitable to avoid these situations


Possible solutions 

- use Pointer field
- use `+kubebuilder:validation:Minimum=8`

cc @tjungblu @dusk125 @hasbro17 @deads2k @soltysh

Reply:
see https://github.com/openshift/api/pull/1736#issuecomment-2069623428 for details

Reply:
first time that the integration passed 🟢  [link](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_api/1736/pull-ci-openshift-api-master-integration/1782424939345743872)

I used these tags, can u please update the enhancement accordingly 


```
        // BackendQuotaGiB sets the etcd backend storage size limit in gibibytes. Defaults to 8GiB.
	// +kubebuilder:validation:Optional
	// +kubebuilder:validation:Minimum=8
	// +kubebuilder:validation:Maximum=32
	// +kubebuilder:validation:XValidation:rule="self>=oldSelf",message="can't decrease database size"
	// +openshift:enable:FeatureGate=EtcdBackendQuota
	// +optional
```